### ISSUE TITLE
coverageByTranscript weights ?

### ISSUE THREAD

**Issue by Roleren** [Bioconductor/GenomicFeatures#19]
For human data, using human NGS reads (1.4GB) as GRanges:

# Using reduced data that has a score column for duplicated reads, every unique read is only once.
system.time(withWeights <- coverageByTranscript(a, cds, weight = "score"))
   user  system elapsed 
  3.987   0.683   4.667 
# Using raw data, without reducing the data-set by not creating a $score column for number of duplicated reads
> system.time(withoutWeights <- coverageByTranscript(b, cds))
   user  system elapsed 
 37.956  21.586  59.547 

I just changed the coverageByTranscript function to allow weights, maybe a possible improvement?

**Comment by Miswi** [comment]
The reason why GenomicFeatures developers didn't allow for weight in coverageByTranscript() is probably coverage method on GrangesList behavior. I elaborate on that in ORFik issues:
https://github.com/Roleren/ORFik/issues/98

**Comment by Roleren** [comment]
I have this implemented as coverageByTranscriptW in ORFik for GRanges input. 
[](https://github.com/Roleren/ORFik/blob/97ed5bda8eaaf9cc989b6a0b02533bb665220488/R/grl_helpers.R#L446)

**Comment by lshep** [comment]
If you would like to make a pull request for the enhancement that would be fantastic.

**Comment by Roleren** [comment]
It should work to merge directly, I will add some more tests, and also add a note to remove the part:
```
 if ((is(x, "GAlignments") | is(x, "GAlignmentPairs"))
      & is.character(weight)) {
```
When coverage(), is fixed for GRangesList.

**Comment by lshep** [comment]
That would be great and when ready please put in the pull request

**Comment by Roleren** [comment]
As a note for you guys, I think coverage(GRangesList) should be changed to support meta columns asap, then much of my function is no longer needed.

**Comment by Roleren** [comment]
Pull request: 
https://github.com/Bioconductor/GenomicFeatures/pull/24

**Comment by Roleren** [comment]
A new update, reading about other people also noting that S4Vectors::revElements is slow, I checked with profvis, and indeed this is the slow part of transcriptByCoverage:

I added a simple change in ORFik:::transcriptByCoverageW to make it much faster:

```
library(S4Vectors)
#' Faster revElements
#' @param x RleList
#' @return RleList (revered list elements)
revElementsF <- function(x) {
  b <- rev(x)
  b@unlistData <- rev(x@unlistData)
  return(rev(b))
}

# Make a large RleList of type integer-Rle
out <- sample(c(0L,1L,5L), size = 100000000, replace = T, prob = c(0.95, 0.4, 0.1))
out <- split(out, sort(sample(seq(length(out)/10), size = length(out), replace = T)))
out <- RleList(out)
out
system.time(a <- revElements(out))
system.time(b <- revElementsF(out))
identical(a, b)

# Adding longer names makes revElements even worse
names(out) <- paste0("ENST", sample(seq(100000, 100000 + length(out)), size = length(out), replace = T))
head(names(out))
system.time(a <- revElements(out))
system.time(b <- revElementsF(out))
identical(a, b)
```
